mesh/mesh_connections.hpp

```diff
@@ -0,0 +1,34 @@
+class MeshConnections
+{
+   private:
+   Mesh *mesh;
+   Array<Array<Table*>> T;
+
+   public:
+   struct EntityID
+   {
+      int dim;
+      int id;
+      bool boundary;
+   }
+
+   struct EntityIDs
+   {
+      int dim;
+      Array<int> ids;
+      bool boundary;
+   }
+
+   //Dimension independent interface
+   bool        IsChild(EntityID parent, EntityID child);
+   Array<int> &Neighbors(EntityID entity, int shared_dim);
+   Array<int> &ChildrenOfEntity(EntityID parent, int child_dim);
+   Array<int> &ChildrenOfEntities(EntityIDs parents, int child_dim);
+   Array<int> &ParentsOfEntity(EntityID child, int parent_dim);
+   Array<int> &ParentsOfAnyEntities(EntityIDs children, int parent_dim);
+   Array<int> &ParentsCoveredByEntities(EntityIDs children, int parent_dim);
```

Comment:
Sorry, I am a bit late to the party. If I understand correctly, then this is going to replace the internal topology interface in the mesh?

Reading trough the existing discussion the naming makes sense to me. However, I think the names `parent` and `children` might be confusing when keeping in mind that we also have NC meshes. Not super sure if this really could be an issue.

I also want to quickly comment on neighborhoods, since I had multiple discussion on this topic in my group, "neighbors" can be ambiguous in non-simplical elements. Let us take for example vertex neighborhoods in quadrilaterals on vertex 2 within the following image
```
1--2--3
|  |  |
3--4--5
```
If we define neighborhoods over the edges, then the neighborhood would be {1,3,4}. If we define the neighborhood over elements, then we obtain {1,2,3,4,5,6}. Again, not super sure how problematic this is in practice, but I just wanted to expose this possible issue, since I faced it in my group.

Reply:
Good points. I think your point about the neighbors is addressed in my comment here: https://github.com/mfem/mfem/pull/2951#discussion_r840772947.

The point is that the neighborhood function would take as a parameter the "shared entity dimension", i.e. you could ask for all "edge neighbor vertices of vertex 2" or all "element neighbor vertices of vertex 2" as different queries.

@acfisher, I think the `Neighbors` function should have an additional `neighbor_dim` argument.

Reply:
Ah, yes, you are right. I read it and somehow managed to forget it when I arrived at writing the comment, sorry.

Reply:
Yeah, I realize now that I had been assuming entity_dim+1 for vertices and entity_dim-1 for other entities.  I like the idea of just naming the dimension you neighboring across.